torch/csrc/jit/frontend/ir_emitter.h

```diff
@@ -1,21 +1,319 @@
 #pragma once
+
 #include <functional>
 #include <memory>
 #include <string>
 
-#include <torch/csrc/jit/api/module.h>
-#include <torch/csrc/jit/frontend/error_report.h>
-#include <torch/csrc/jit/frontend/resolver.h>
-#include <torch/csrc/jit/frontend/sugared_value.h>
-#include <torch/csrc/jit/frontend/tree_views.h>
+#include <torch/csrc/jit/frontend/canonicalize_modified_loop.h>
+#include <torch/csrc/jit/frontend/convert_to_ssa.h>
+#include <torch/csrc/jit/frontend/environment.h>
+#include <torch/csrc/jit/frontend/parser.h>
+#include <torch/csrc/jit/frontend/run_cleanup_passes.h>
+#include <torch/csrc/jit/frontend/script_type_parser.h>
+#include <torch/csrc/jit/frontend/type_refinement.h>
+
 #include <torch/csrc/jit/ir/ir.h>
 
+#include <torch/csrc/jit/passes/normalize_ops.h>
+
 namespace torch {
 namespace jit {
 
-TORCH_API void runCleanupPasses(std::shared_ptr<Graph>& to_clean);
```

Comment:
Do all of the functions, aliases, and types declared here need to be exposed? On the master branch, this file declares only two functions: `runCleanupPasses` and `meaningfulName`. Are all of these other declarations that have been added used by files that include this header?

If not, consider moving them into the `.cpp`.

Reply:
I can move the `using` declarations, etc. into the `.cpp` file, but I do need to declare some of the internal methods of `to_ir` in the header file. So, I thought it would be best to declare all the internal methods in the header file. I was under the impression that best practice was keeping the "API reference" in a single location (either the header OR the source file). Is that only applicable in certain circumstances?

Reply:
That's true, but most of the stuff in the header is not a part of the API of `to_ir`. Unfortunately, it is not possible to split the declaration of a class into multiple pieces (like you can do for a namespace). Move what you can, but it looks like a lot of the `emitXXX` methods use the `graph` or `environment` member of `to_ir`, so they cannot be easily converted into anonymous functions that live only in the `.cpp`. We could consider using the [`pImpl` idiom](https://en.cppreference.com/w/cpp/language/pimpl) to hide these extra details, but I think that should be in a future PR.